File: lib/core/reader/locator/reader_locator_codec.dart
class ReaderLocatorCodec {
  static const _epubScrollPrefix = 'epub-scroll:';
  static const _pdfPagePrefix = 'pdf-page:';

  static String encodeEpubScrollPercent(double percent) {
    return '$_epubScrollPrefix${percent.clamp(0.0, 1.0)}';
  }

  static double? decodeEpubScrollPercent(String? locator) {
    if (locator == null || !locator.startsWith(_epubScrollPrefix)) return null;
    final value = double.tryParse(locator.substring(_epubScrollPrefix.length));
    return value?.clamp(0.0, 1.0);
  }

  static String encodePdfPageIndex(int pageIndex) {
    return '$_pdfPagePrefix${pageIndex < 0 ? 0 : pageIndex}';
  }

  static int? decodePdfPageIndex(String? locator) {
    if (locator == null || !locator.startsWith(_pdfPagePrefix)) return null;
    final value = int.tryParse(locator.substring(_pdfPagePrefix.length));
    if (value == null) return null;
    return value < 0 ? 0 : value;
  }
}

File: core/reader/src/main/kotlin/com/readler/core/reader/locator/ReaderLocatorCodec.kt
package com.readler.core.reader.locator

object ReaderLocatorCodec {
    private const val EPUB_SCROLL_PREFIX = "epub-scroll:"
    private const val PDF_PAGE_PREFIX = "pdf-page:"

    fun encodeEpubScrollPercent(percent: Float): String {
        return "$EPUB_SCROLL_PREFIX${percent.coerceIn(0f, 1f)}"
    }

    fun decodeEpubScrollPercent(locator: String?): Float? {
        if (locator == null || !locator.startsWith(EPUB_SCROLL_PREFIX)) {
            return null
        }

        return locator.removePrefix(EPUB_SCROLL_PREFIX).toFloatOrNull()?.coerceIn(0f, 1f)
    }

    fun encodePdfPageIndex(pageIndex: Int): String {
        return "$PDF_PAGE_PREFIX${pageIndex.coerceAtLeast(0)}"
    }

    fun decodePdfPageIndex(locator: String?): Int? {
        if (locator == null || !locator.startsWith(PDF_PAGE_PREFIX)) {
            return null
        }

        return locator.removePrefix(PDF_PAGE_PREFIX).toIntOrNull()?.coerceAtLeast(0)
    }
}
